### ISSUE TITLE
SIR model with exponential waiting times in Julia

### ISSUE THREAD

**Issue by mschauer** [epirecipes/sir-julia#35]
How do you like something very simple, I think you don't have that yet:

https://gist.github.com/mschauer/be1196dedbf8afe7fb8a0b411672b4d6

**Comment by slwu89** [comment]
hi @mschauer, it looks like there are a few CTMC constructions of the SIR model in this repo, just listed as "jump processes":
1. https://github.com/epirecipes/sir-julia/blob/master/markdown/jump_process/jump_process.md
2. https://github.com/epirecipes/sir-julia/blob/master/markdown/jump_process_gillespie/jump_process_gillespie.md

**Comment by mschauer** [comment]
Think of it in the same fashion as https://discourse.julialang.org/t/improving-metropolis-hastings-code/67412/7, you could use Turing there too instead.

**Comment by sdwfrost** [comment]
I think this is already covered for now, unless @mschauer, you think there's something we're missing with the jump process implementations

**Comment by mschauer** [comment]
> unless you think there's something we're missing with the jump process implementations

I would think so:
```
prob_jump = JumpProblem(prob,Direct(),infection_jump,recovery_jump);
sol_jump = solve(prob_jump,SSAStepper());
 ```
 has a very different didactical value compared to
 ```
 function sir!(us, u, T, p)
    β, c, γ = p.β, p.c, p.γ
    while true
        t, S, I, R = u
       # @show t, u
        N = +(S, I, R)
        τ = randexp()/(S * β*c*I/N), randexp()/(I*γ) # renew both times, because state has changed
        t + min(τ[1], τ[2]) > T && return us
        if τ[1] < τ[2]
            t += τ[1]
            S -= 1
            I += 1
        else
            t += τ[2]
            I -= 1
            R += 1
        end
        u = (t, S, I, R)
        push!(us, u)
    end
end
```

**Comment by sdwfrost** [comment]
OK, I can add the plain Julia code, as I agree that there are didactic differences. This no-library approach was in the [old repo](http://epirecip.es/epicookbook/chapters/sir-stochastic-discretestate-continuoustime/julia), and was in an [old blog post of mine](http://phylodynamics.blogspot.com/2013/06/an-sir-model-in-julia.html) too

**Comment by slwu89** [comment]
@sdwfrost I think if we are going to add plain Julia code, do you think its worth it to do plain Julia for different flavors of how to sample the CTMC? @mschauer has posted a nice Direct method, it might be worth it to show a First Reaction Method and Anderson's Next Reaction Method [paper here](https://people.math.wisc.edu/~anderson/papers/AndNRM.pdf).

**Comment by mschauer** [comment]
I think have something close to *Algorithm 2 (the next reaction method)* from the Anderson paper, except that I simply resample both times instead what is done in step (8) to reuse elements of the vector of random next event times.

**Comment by sdwfrost** [comment]
@slwu89 you probably have something along these lines hanging around too?

**Comment by slwu89** [comment]
Yeah, I'd love to write up an example with a direct method, first reaction method, and Anderson's NRM. I think I also coded [Thanh's rejection sampler](https://aip.scitation.org/doi/abs/10.1063/1.4896985?ver=pdfcov&journalCode=jcp) in R before.

**Comment by sdwfrost** [comment]
Hi @slwu89 - just seeing if you're still interested in this!
